lib/enigma.rb

```diff
@@ -111,38 +108,56 @@ def find_key(encrypted_message, date)
     key  = ""
     offsets = date_to_offsets(date)
     find_shifts(encrypted_message, date).each_with_index do |shift, index|
-      if ((shift - offsets[index]) % 27).to_s.length == 1 && index == 0
-        key.concat("0" + ((shift - offsets[index]) % 27).to_s)
-      elsif index == 0
-        key.concat(((shift - offsets[index]) % 27).to_s)
+      if index == 0
+        key.concat(find_first_key(shift, offsets[index]))
       else
-        new_key_value = 0
-        until key.length > index + 1
-          if (key.chars.last.concat(new_key_value.to_s)).to_i % 27 == shift
-            key.concat((((new_key_value - offsets[index]) % 27).to_s))
-          end
-          new_key_value += 1
-        end
+        find_next_key_value(key, index, offsets[index], shift)
       end
     end
     key
   end
 
+  def find_first_key(shift, offset)
+    if ((shift - offset) % 27).to_s.length == 1
+      "0" + ((shift - offset) % 27).to_s
+    else
+      ((shift - offset) % 27).to_s
+    end
+  end
+
+  def find_next_key_value(key, index, offset, shift)
+    new_key_value = 0
+    until key.length > index + 1
+      if (key.chars.last.concat(new_key_value.to_s)).to_i % 27 == shift
+        key.concat((((new_key_value - offset) % 27).to_s))
+      end
+      new_key_value += 1
+    end
+  end
+
```

Comment:
Refactored this into helper methods. I don't think this is the problem for finding the key even though the shifts will be smaller than the actual shifts with the key if the original is greater than 27.